File: malai-core/org.malai.core/src/main/org/malai/binding/WidgetBindingImpl.java
/*
 * This file is part of Malai.
 * Copyright (c) 2005-2017 Arnaud BLOUIN
 * Malai is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * Malai is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package org.malai.binding;

import java.util.Arrays;
import java.util.logging.Level;
import java.util.logging.Logger;
import javafx.beans.property.Property;
import org.malai.action.Action;
import org.malai.action.ActionImpl;
import org.malai.action.ActionsRegistry;
import org.malai.action.AutoUnbind;
import org.malai.error.ErrorCatcher;
import org.malai.instrument.Instrument;
import org.malai.interaction.Interaction;
import org.malai.stateMachine.MustCancelStateMachineException;
import org.malai.undo.Undoable;

/**
 * The base class to do widget bindings, i.e. bindings between user interactions and (undoable) actions.
 * @param <A> The type of the action that will produce this widget binding.
 * @param <I> The type of the interaction that will use this widget binding.
 * @param <N> The type of the instrument that will contain this widget binding.
 * @author Arnaud BLOUIN
 */
public abstract class WidgetBindingImpl<A extends ActionImpl, I extends Interaction, N extends Instrument<?>> implements WidgetBinding {
	protected Logger loggerBinding;

	protected Logger loggerAction;

	/** The source interaction. */
	protected final I interaction;

	/** The current action in progress. */
	protected A action;

	/** The instrument that contains the widget binding. */
	protected final N instrument;

	/** Specifies if the action must be execute or update * on each evolution of the interaction. */
	protected final boolean execute;

	/** Defines whether the action must be executed in a specific thread. */
	protected boolean async;

	/** The action class to instantiate. */
	protected final Class<A> clazzAction;


	/**
	 * Creates a widget binding. This constructor must initialise the interaction. The widget binding is (de-)activated if the given
	 * instrument is (de-)activated.
	 * @param ins The instrument that contains the widget binding.
	 * @param exec Specifies if the action must be execute or update on each evolution of the interaction.
	 * @param actionClass The type of the action that will be created. Used to instantiate the action by reflexivity.
	 * The class must be public and must have a constructor with no parameter.
	 * @param interaction The user interaction of the binding.
	 * @throws IllegalAccessException If no free-parameter constructor is available.
	 * @throws InstantiationException If an error occurs during instantiation of the interaction/action.
	 * @throws IllegalArgumentException If the given interaction or instrument is null.
	 */
	public WidgetBindingImpl(final N ins, final boolean exec, final Class<A> actionClass, final I interaction) throws InstantiationException, IllegalAccessException {
		super();

		if(ins == null || actionClass == null || interaction == null) {
			throw new IllegalArgumentException();
		}

		clazzAction = actionClass;
		this.interaction = interaction;
		action = null;
		instrument = ins;
		execute = exec;
		this.interaction.addHandler(this);
		setActivated(ins.isActivated());
		async = false;
	}

	public void logBinding(final boolean log) {
		if(log) {
			if(loggerBinding == null) {
				loggerBinding = Logger.getLogger(getClass().getName());
			}
		}else {
			loggerBinding = null;
		}
	}

	public void logAction(final boolean log) {
		if(log) {
			if(loggerAction == null) {
				loggerAction = Logger.getLogger(getClass().getName());
			}
		}else {
			loggerAction = null;
		}
	}

	public void logInteraction(final boolean log) {
		interaction.log(log);
	}

	/**
	 * Whether the action must be executed in a specific thread.
	 * @return True: the action will be executed asynchronously.
	 */
	public boolean isAsync() {
		return async;
	}

	/**
	 * Sets wether the action must be executed in a specific thread.
	 * @param asyncAction True: the action will be executed asynchronously.
	 */
	public void setAsync(final boolean asyncAction) {
		async = asyncAction;
	}

	@Override
	public void clearEvents() {
		interaction.clearEvents();
	}


	/**
	 * creates the action of the widget binding. If the attribute 'action' is not null, nothing will be done.
	 * @return The created action or null if problems occured.
	 */
	protected A createAction() {
		try {
			return clazzAction.newInstance();
		}catch(final IllegalAccessException | InstantiationException ex) {
			ErrorCatcher.INSTANCE.reportError(ex);
			return null;
		}
	}


	@Override
	public abstract void initAction();


	@Override
	public void updateAction() {
		// to override.
	}


	@Override
	public abstract boolean isConditionRespected();


	@Override
	public I getInteraction() {
		return interaction;
	}


	@Override
	public A getAction() {
		return action;
	}


	@Override
	public boolean isActivated() {
		return instrument.isActivated();
	}

	@Override
	public boolean isRunning() {
		return interaction.isRunning();
	}


	@Override
	public boolean isStrictStart() {
		return false;
	}

	protected void unbindActionAttributes() {
		if(action != null) {
			unbindActionAttributesClass(action.getClass());
			if(loggerAction != null) {
				loggerAction.log(Level.INFO, "Action unbound: " + action);
			}
		}
	}

	private void unbindActionAttributesClass(final Class<? extends ActionImpl> clazz) {
		Arrays.stream(clazz.getDeclaredFields()).filter(field -> field.isAnnotationPresent(AutoUnbind.class) &&
			Property.class.isAssignableFrom(field.getType())).forEach(field -> {
			try {
				final boolean access = field.isAccessible();
				field.setAccessible(true);
				final Object o = field.get(action);
				if(o instanceof Property<?>) {
					((Property<?>) o).unbind();
				}
				field.setAccessible(access);
			}catch(final IllegalAccessException ex) {
				ex.printStackTrace();
			}
		});

		final Class<?> superClass = clazz.getSuperclass();
		if(superClass != null && superClass != ActionImpl.class && ActionImpl.class.isAssignableFrom(superClass)) {
			unbindActionAttributesClass((Class<? extends ActionImpl>) superClass);
		}
	}

	@Override
	public void interactionCancels() {
		if(action != null) {
			if(loggerBinding != null) {
				loggerBinding.log(Level.INFO, "Binding cancelled");
			}

			action.cancel();
			if(loggerAction != null) {
				loggerAction.log(Level.INFO, "Action cancelled");
			}
			unbindActionAttributes();

			// The instrument is notified about the cancel of the action.
			instrument.onActionCancelled(action);

			if(isExecute()) {
				if(action instanceof Undoable) {
					((Undoable) action).undo();
					if(loggerAction != null) {
						loggerAction.log(Level.INFO, "Action undone");
					}
				}else {
					throw new MustBeUndoableActionException(action.getClass());
				}
			}

			action = null;
			instrument.interimFeedback();
		}
	}


	@Override
	public void interactionStarts() throws MustCancelStateMachineException {
		final boolean ok = action == null && isActivated() && isConditionRespected();

		if(loggerBinding != null) {
			loggerBinding.log(Level.INFO, "Starting binding: " + ok);
		}

		if(ok) {
			action = createAction();
			initAction();
			if(loggerAction != null) {
				loggerAction.log(Level.INFO, "Action created and init: " + action);
			}
			interimFeedback();
		}else {
			if(isStrictStart()) {
				if(loggerBinding != null) {
					loggerBinding.log(Level.INFO, "Cancelling starting interaction: " + interaction);
				}
				throw new MustCancelStateMachineException();
			}
		}
	}


	@Override
	public void interactionStops() {
		final boolean ok = isConditionRespected();
		if(loggerBinding != null) {
			loggerBinding.log(Level.INFO, "Binding stops with condition: " + ok);
		}
		if(ok) {
			if(action == null) {
				action = createAction();
				initAction();
				if(loggerAction != null) {
					loggerAction.log(Level.INFO, "Action created and init: " + action);
				}
			}

			if(!execute) {
				updateAction();
				if(loggerAction != null) {
					loggerAction.log(Level.INFO, "Action updated: " + action);
				}
			}

			executeAction(action);
			unbindActionAttributes();
			action = null;
			instrument.interimFeedback();
		}else {
			if(action != null) {
				if(loggerAction != null) {
					loggerAction.log(Level.INFO, "Cancelling the action: " + action);
				}
				action.cancel();
				unbindActionAttributes();
				instrument.onActionCancelled(action);
				action = null;
				instrument.interimFeedback();
			}
		}
	}


	private void executeAction(final Action act) {
		boolean ok = act.doIt();
		if(loggerAction != null) {
			loggerAction.log(Level.INFO, "Action execution did it: " + ok);
		}

		if(ok) {
			instrument.onActionExecuted(act);
			act.done();
			instrument.onActionDone(act);
		}

		ok = act.hadEffect();

		if(loggerAction != null) {
			loggerAction.log(Level.INFO, "Action execution had effect: " + ok);
		}
		if(ok) {
			if(act.getRegistrationPolicy() != Action.RegistrationPolicy.NONE) {
				ActionsRegistry.INSTANCE.addAction(act, instrument);
				instrument.onActionAdded(act);
			}else {
				ActionsRegistry.INSTANCE.unregisterActions(act);
			}
			act.followingActions().forEach(this::executeAction);
		}
	}


	@Override
	public void interactionUpdates() {
		final boolean ok = isConditionRespected();

		if(loggerBinding != null) {
			loggerBinding.log(Level.INFO, "Binding updates with condition: " + ok);
		}

		if(ok) {
			if(action == null) {
				if(loggerAction != null) {
					loggerAction.log(Level.INFO, "Action creation");
				}
				action = createAction();
				initAction();
			}

			if(loggerAction != null) {
				loggerAction.log(Level.INFO, "Action update");
			}

			updateAction();

			if(execute && action.canDo()) {
				if(loggerAction != null) {
					loggerAction.log(Level.INFO, "Action execution");
				}
				action.doIt();
				instrument.onActionExecuted(action);
			}

			interimFeedback();
		}
	}


	@Override
	public boolean isExecute() {
		return execute;
	}


	@Override
	public void interimFeedback() {
		//
	}


	@Override
	public void setActivated(final boolean activ) {
		if(loggerBinding != null) {
			loggerBinding.log(Level.INFO, "Binding Activated: " + activ);
		}

		interaction.setActivated(activ);
	}


	@Override
	public N getInstrument() {
		return instrument;
	}
}

File: malai-core/org.malai.ts/src/src-core/binding/WidgetBindingImpl.ts
/*
 * This file is part of Malai.
 * Copyright (c) 2009-2018 Arnaud BLOUIN
 * Malai is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * Malai is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

import {InteractionImpl} from "../interaction/InteractionImpl";
import {WidgetBinding} from "./WidgetBinding";
import {CancelFSMException} from "../fsm/CancelFSMException";
import {isUndoableType} from "../undo/Undoable";
import {factory} from "../logging/ConfigLog";
import {Logger} from "typescript-logging";
import {FSM} from "../fsm/FSM";
import {CommandImpl} from "../command/CommandImpl";
import {MustBeUndoableCmdException} from "./MustBeUndoableCmdException";
import {Command, RegistrationPolicy} from "../command/Command";
import {CommandsRegistry} from "../command/CommandsRegistry";
import {InteractionData} from "../interaction/InteractionData";

/**
 * Creates a widget binding. This constructor must initialise the interaction. The widget binding is (de-)activated if the given
 * instrument is (de-)activated.
 * @param {*} ins The instrument that contains the widget binding.
 * @param {boolean} exec Specifies if the command must be execute or update on each evolution of the interaction.
 * @param {*} actionClass The type of the command that will be created. Used to instantiate the command by reflexivity.
 * The class must be public and must have a constructor with no parameter.
 * @param {InteractionImpl} interaction The user interaction of the binding.
 * @throws IllegalArgumentException If the given interaction or instrument is null.
 * @class
 * @author Arnaud BLOUIN
 */
export abstract class WidgetBindingImpl<C extends CommandImpl, I extends InteractionImpl<D, {}, FSM<{}>>, D extends InteractionData>
            implements WidgetBinding {
    protected loggerBinding: Logger | undefined;

    protected loggerCmd: Logger | undefined;

    /**
     * The source interaction.
     */
    protected readonly interaction: I;

    /**
     * The current action in progress.
     */
    protected cmd: C | undefined;

    /**
     * Specifies if the command must be execute or update * on each evolution of the interaction.
     */
    protected execute: boolean;

    /**
     * Defines whether the command must be executed in a specific thread.
     */
    protected async: boolean;

    /**
     * The command class to instantiate.
     */
    private readonly cmdProducer: (i: D) => C;

    protected constructor(exec: boolean, interaction: I, cmdClass: (i: D) => C) {
        this.execute = false;
        this.async = false;
        this.cmdProducer = cmdClass;
        this.interaction = interaction;
        this.cmd = undefined;
        // this.instrument = ins;
        this.execute = exec;
        this.interaction.getFsm().addHandler(this);
        this.setActivated(true); //ins.isActivated());
    }

    public logBinding(log: boolean): void {
        if (log) {
            if (this.loggerBinding === undefined) {
                this.loggerBinding = factory.getLogger("Binding");
            }
        } else {
            //     this.loggerBinding = undefined;
        }
    }

    public logCmd(log: boolean): void {
        if (log) {
            if (this.loggerCmd === undefined) {
                this.loggerCmd = factory.getLogger("Command");
            }
        } else {
            this.loggerCmd = undefined;
        }
    }

    public logInteraction(log: boolean): void {
        this.interaction.log(log);
    }

    /**
     * Whether the command must be executed in a specific thread.
     * @return {boolean} True: the command will be executed asynchronously.
     */
    public isAsync(): boolean {
        return this.async;
    }

    /**
     * Sets wether the command must be executed in a specific thread.
     * @param {boolean} asyncCmd True: the command will be executed asynchronously.
     */
    public setAsync(asyncCmd: boolean) {
        this.async = asyncCmd;
    }

    /**
     *
     */
    public clearEvents(): void {
        this.interaction.fullReinit();
    }

    /**
     * creates the command of the widget binding. If the attribute 'cmd' is not null, nothing will be done.
     * @return {CommandImpl} The created command.
     */
    protected map(): C {
        return this.cmdProducer(this.interaction.getData());
    }

    public abstract first(): void;

    public then(): void {
    }

    /**
     * @return {boolean}
     */
    public abstract when(): boolean;

    public getInteraction(): I {
        return this.interaction;
    }

    /**
     * @return {CommandImpl}
     */
    public getCommand(): C | undefined {
        return this.cmd;
    }

    /**
     * @return {boolean}
     */
    public isActivated(): boolean {
        return true; //this.instrument.isActivated();
    }

    /**
     * @return {boolean}
     */
    public isRunning(): boolean {
        return this.interaction.isRunning();
    }

    /**
     *
     * @return {boolean}
     */
    public isStrictStart(): boolean {
        return false;
    }

    protected unbindCmdAttributes(): void {
        if (this.cmd !== undefined) {
            this.unbindCmdAttributesClass(this.cmd.constructor);
            if (this.loggerCmd !== undefined) {
                this.loggerCmd.info("Command unbound: " + String(this.cmd));
            }
        }
    }

    private unbindCmdAttributesClass(clazz: Object): void {
        //FIXME
        // java.util.Arrays.stream<any>(clazz.getDeclaredFields()).filter((field) =>
        // field.isAnnotationPresent("AutoUnbind") && "javafx.beans.property.Property".isAssignableFrom(field.getType())).
        // forEach((field) => {
        //     try {
        //         let access : boolean = field.isAccessible();
        //         let o : any = /* get */this.cmd[field.name];
        //         if(o instanceof Property) {
        //             (<javafx.beans.property.Property<any>><any>o).unbind();
        //         }
        //     } catch(ex) {
        //         console.error(ex.message, ex);
        //     };
        // });
        // let superClass : any = clazz.getSuperclass();
        // if(superClass !== undefined && superClass !== ActionImpl && ActionImpl.isAssignableFrom(superClass)) {
        //     this.unbindActionAttributesClass(<any>superClass);
        // }
    }

    /**
     *
     */
    public fsmCancels(): void {
        if (this.cmd !== undefined) {
            if (this.loggerBinding !== undefined) {
                this.loggerBinding.info("Binding cancelled");
            }
            this.cmd.cancel();
            if (this.loggerCmd !== undefined) {
                this.loggerCmd.info("Command cancelled");
            }
            this.unbindCmdAttributes();
            // this.instrument.onCmdCancelled(this.cmd);
            if (this.isExecute() && this.cmd.hadEffect()) {
                if (isUndoableType(this.cmd)) {
                    this.cmd.undo();
                    if (this.loggerCmd !== undefined) {
                        this.loggerCmd.info("Command undone");
                    }
                } else {
                    throw new MustBeUndoableCmdException(this.cmd);
                }
            }
            this.cmd = undefined;
            // this.instrument.interimFeedback();
        }
    }

    /**
     *
     */
    public fsmStarts(): void {
        const ok: boolean = this.cmd === undefined && this.isActivated() && this.when();
        if (this.loggerBinding !== undefined) {
            this.loggerBinding.info("Starting binding: " + String(ok));
        }
        if (ok) {
            this.cmd = this.map();
            this.first();
            if (this.loggerCmd !== undefined) {
                this.loggerCmd.info("Command created and init: " + String(this.cmd));
            }
            this.feedback();
        } else {
            if (this.isStrictStart()) {
                if (this.loggerBinding !== undefined) {
                    this.loggerBinding.info("Cancelling starting interaction: " + String(this.interaction));
                }
                throw new CancelFSMException();
            }
        }
    }

    /**
     *
     */
    public fsmStops(): void {
        const ok: boolean = this.when();
        if (this.loggerBinding !== undefined) {
            this.loggerBinding.info("Binding stops with condition: " + String(ok));
        }
        if (ok) {
            if (this.cmd === undefined) {
                this.cmd = this.map();
                this.first();
                if (this.loggerCmd !== undefined) {
                    this.loggerCmd.info("Command created and init: " + String(this.cmd));
                }
            }
            if (!this.execute) {
                this.then();
                if (this.loggerCmd !== undefined) {
                    this.loggerCmd.info("Command updated: " + String(this.cmd));
                }
            }
            this.executeCmd(this.cmd, this.async);
            this.unbindCmdAttributes();
            this.cmd = undefined;
            // this.instrument.interimFeedback();
        } else {
            if (this.cmd !== undefined) {
                if (this.loggerCmd !== undefined) {
                    this.loggerCmd.info("Cancelling the command: " + String(this.cmd));
                }
                this.cmd.cancel();
                this.unbindCmdAttributes();
                // this.instrument.onCmdCancelled(this.cmd);
                this.cmd = undefined;
                // this.instrument.interimFeedback();
            }
        }
    }

    private executeCmd(cmd: Command, async: boolean): void {
        if (async) {
            this.executeCmdAsync(cmd);
        } else {
            this.afterCmdExecuted(cmd, cmd.doIt());
        }
    }

    protected abstract executeCmdAsync(cmd: Command): void;

    protected afterCmdExecuted(cmd: Command, ok: boolean): void {
        if (this.loggerCmd !== undefined) {
            this.loggerCmd.info("Command execution did it: " + String(ok));
        }
        if (ok) {
            // this.instrument.onCmdExecuted(cmd);
            cmd.done();
            // this.instrument.onCmdDone(cmd);
        }
        const hadEffect: boolean = cmd.hadEffect();
        if (this.loggerCmd !== undefined) {
            this.loggerCmd.info("Command execution had effect: " + String(hadEffect));
        }
        if (hadEffect) {
            if (cmd.getRegistrationPolicy() !== RegistrationPolicy.NONE) {
                CommandsRegistry.INSTANCE.addCommand(cmd); //, this.instrument
                // this.instrument.onCmdAdded(cmd);
            } else {
                CommandsRegistry.INSTANCE.unregisterCmd(cmd);
            }
            cmd.followingCmds().forEach(cmdFollow => this.executeCmd(cmdFollow, false));
        }
    }

    public fsmUpdates(): void {
        const ok: boolean = this.when();
        if (this.loggerBinding !== undefined) {
            this.loggerBinding.info("Binding updates with condition: " + String(ok));
        }
        if (ok) {
            if (this.cmd === undefined) {
                if (this.loggerCmd !== undefined) {
                    this.loggerCmd.info("Command creation");
                }
                this.cmd = this.map();
                this.first();
            }
            if (this.loggerCmd !== undefined) {
                this.loggerCmd.info("Command update");
            }
            this.then();
            if (this.execute && this.cmd.canDo()) {
                if (this.loggerCmd !== undefined) {
                    this.loggerCmd.info("Command execution");
                }
                this.cmd.doIt();
                // this.instrument.onCmdExecuted(this.cmd);
            }
            this.feedback();
        }
    }

    public uninstallBinding(): void {
        this.setActivated(false);
        this.loggerCmd = undefined;
        this.loggerBinding = undefined;
    }

    /**
     *
     * @return {boolean}
     */
    public isExecute(): boolean {
        return this.execute;
    }

    /**
     *
     */
    public feedback(): void {
    }

    /**
     *
     * @param {boolean} activ
     */
    public setActivated(activ: boolean): void {
        if (this.loggerBinding !== undefined) {
            this.loggerBinding.info("Binding Activated: " + String(activ));
        }
        this.interaction.setActivated(activ);
    }
}
